### DISCUSSION TITLE
Image Function Capabilities

### DISCUSSION THREAD

**Discussion by agorgogi** [idaholab/moose#26733] [Q&A Meshing] [answered]
Hello,
I would like to use Moose for simulating heterogeneous materials. I have input images of microstructures, an example is attached, which I want to mesh with Moose so as to accurately capture the geometry of the phases, and create different block IDs for the finite elements belonging to a different phase ( each phase is shown with different color in the image). I could then refer to finite elements based on their block ID, and assign the corresponding phase material properties to them. 
![0000](https://github.com/idaholab/moose/assets/94208556/1f747202-da91-45c8-8829-a0e417416cde)

My first approach was the following: 
Create a uniform mesh with finite element size matching the pixel/voxel size of the image, and then passing this mesh as the input to ImageSubdomain Generator, as follows

```
[Mesh]
    [mesh_define]
        type = GeneratedMeshGenerator
        dim = 2
        nx = 320 # = number of image pixels in x
        ny = 320 # = number of image pixels in y
    []
    [image]
        input = mesh_define
        type = ImageSubdomainGenerator
        file_base = microstructure396_3D_png/0  # all images filenames start with 0
        file_suffix = png
        file_range = '0000' # read only one image from the stack for now
        threshold = 255 # everything above 255 will be assigned to block 1, everything below to block 0
        upper_value = 1
        lower_value = 0
    []  
[]
```

This results in accurate geometry representation and assignment of block IDs to the different phases, as shown below.
![image](https://github.com/idaholab/moose/assets/94208556/ced75b65-7819-460f-8762-76325e869c9e)
The problem with this approach is that it is very computationally expensive, because of the uniform mesh. 

So my second approach was to use adaptive mesh refinement with Moose ImageFunction. Moose image function will sample the values of the image (the values of the different pixel colors), and will adaptively refine the mesh until it adequetely captures the geometrical features in the image. In the Moose documentation, it is mentioned that I can then use the result of the Image Function to set the values of a variable, e.g. displacements. However, I do not want that. Instead, what I want is to use the result of the ImageFunction to set the block IDs of the corresponding finite elements. Is there any possible way I could do that? Or is there any other possible way I could create a non-uniform (adaptive) mesh that accurately captures the image geometry, and correctly assigns different block IDs to finite elements belonging to different phases?

```
[Mesh]
    type = GeneratedMesh
    dim = 2
    nx = 10
    ny = 10
[]

[Functions]
    [image_func]
      type = ImageFunction
      file_base = microstructure396_3D_png/0
      file_suffix = png
      file_range = '0000'
      threshold = 255
      upper_value = 1
      lower_value = 0
    []
[]
  
[Variables]
    [u]
    []
[]

[ICs]
    [u_ic]
      type = FunctionIC
      function = image_func
      variable = u
    []
[]
  
[Problem]
    type = FEProblem
    solve = false  # that was the other way no to solve the problem, but only generate the mesh
[]
  
[Executioner]
    type = Transient
    num_steps = 1
    dt = 0.1
[]
  
[Adaptivity]
    max_h_level = 5
    initial_steps = 5
    initial_marker = marker
    [Indicators]
      [indicator]
        type = GradientJumpIndicator
        variable = u
      []
    []
    [Markers]
      [marker]
        type = ErrorFractionMarker
        indicator = indicator
        refine = 0.9
      []
    []
[]
  
[Outputs]
    exodus = true
[]
```

![image](https://github.com/idaholab/moose/assets/94208556/479daa7d-0451-4abd-93a9-ca433fcde2c1)

Looking forward to any suggestions, thank you!

**Comment by GiudGiud** [comment, marked as answer]
Hello

This might be a little convoluted but you could : 
- use a FunctionAux to set an auxiliary variable to the local value of the function (executed on INITIAL)
https://mooseframework.inl.gov/source/auxkernels/FunctionAux.html

- use this user object on INITIAL to change the subdomain
https://mooseframework.inl.gov/source/userobjects/CoupledVarThresholdElementSubdomainModifier.html

**Reply by agorgogi** [reply]
Many thanks! Will look into these options!

**Reply by GiudGiud** [reply]
Max raised a good question. This won't work if we can't have child elements of a same parent element have different subdomains from that parent element. 

I don't know if we can't have that.

**Reply by maxnezdyur** [reply]
@GiudGiud I should have checked before I sent my first comment, which is why I tried to delete it. Relatively new, but it seems like that we actually can have different subdomains. I didn't look too much in the results to see what happens during the run but one of the input file below may do what we want. Not sure how best to link input files.
`test/tests/userobjects/element_subdomain_modifier/adaptivity_moving_boundary.i`

**Reply by GiudGiud** [reply]
Awesome. This should work then.

**Reply by agorgogi** [reply]
@GiudGiud , @maxnezdyur 
Thank you both for the help. Yes, it worked now. All I would have to do is to improve the adaptivity part to better capture the geometrical features of the image, but the block IDs did change as I wanted them to.
![image](https://github.com/idaholab/moose/assets/94208556/1c3ebadd-83e0-4e91-ba52-6d274d1a30ea)


```
[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 10
    ny = 10
  []
  [block0]
    type = SubdomainBoundingBoxGenerator
    input = 'gen'
    block_id = 0
    bottom_left = '0 0 0'
    top_right = '1 0.5 0'
  []
  [block1]
    type = SubdomainBoundingBoxGenerator
    input = 'block0'
    block_id = 1
    bottom_left = '0 0.5 0'
    top_right = '1 1 0'
  []
[]

[Functions]
    [image_func]
      type = ImageFunction
      file_base = microstructure396_3D_png/0
      file_suffix = png
      file_range = '0000'
      threshold = 255
      upper_value = 1
      lower_value = 0
    []
[]
  
[AuxVariables]
  [auxBlockID]
  []
[]

[AuxKernels]
  [aux_kern]
    type = FunctionAux
    variable = auxBlockID
    function = image_func
    execute_on = 'INITIAL'
  []
[]

[UserObjects]
  [change_blockID]
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = auxBlockID
    criterion_type = ABOVE
    threshold = 0
    subdomain_id = 1
    complement_subdomain_id = 0
    execute_on = 'INITIAL'
  []
[]


[Variables]
    [u]
    []
[]

#Try use the auxillary variable in the adaptivity criterion instead of assigning it to the displacement variable
[ICs]
    [u_ic]
      type = FunctionIC
      function = image_func
      variable = u
    []
[]
  
[Problem]
    type = FEProblem
    solve = false  # that was the other way no to solve the problem, but only generate the mesh
[]
  
[Executioner]
    type = Transient
    num_steps = 1
    dt = 0.1
[]
  
[Adaptivity]
    max_h_level = 10
    initial_steps = 5
    initial_marker = marker
    [Indicators]
      [indicator]
        type = GradientJumpIndicator
        variable = u
      []
    []
    [Markers]
      [marker]
        type = ErrorFractionMarker
        indicator = indicator
        refine = 0.9
      []
    []
[]
  
[Outputs]
    exodus = true
[]
```

**Reply by GiudGiud** [reply]
Awesome

**Reply by Rahim-Habibi** [reply]
Hi @agorgogi and @GiudGiud , 
I would like to simulate heterogeneous materials in MOOSE, so after reading the documents "ImageFunction" i have endded up with your issue because i wanted to have several blocks as well. 
Thus, i tried your input file with my own imahe but i did not get the two block, i only have one block see the figure below

![OIP-paraview](https://github.com/user-attachments/assets/52a1529f-f6e5-4f61-8bfe-5c694dbc8c14)

and this the figure i tried to mesh 
![OIP](https://github.com/user-attachments/assets/028265f4-938e-4f5d-aa91-7c323c73d81e)
 


here is the input file :
```
[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 10
    ny = 10
  []
  [block0]
    type = SubdomainBoundingBoxGenerator
    input = 'gen'
    block_id = 0
    bottom_left = '0 0 0'
    top_right = '1 0.5 0'
  []
  [block1]
    type = SubdomainBoundingBoxGenerator
    input = 'block0'
    block_id = 1
    bottom_left = '0 0.5 0'
    top_right = '1 1 0'
  []
[]

[Functions]
    [image_func]
      type = ImageFunction
      file = OIP.png
      threshold = 255
      upper_value = 1
      lower_value = 0
    []
[]

[AuxVariables]
  [auxBlockID]
  []
[]

[AuxKernels]
  [aux_kern]
    type = FunctionAux
    variable = auxBlockID
    function = image_func
    execute_on = 'INITIAL'
  []
[]

[UserObjects]
  [change_blockID]
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = auxBlockID
    criterion_type = ABOVE
    threshold = 0
    subdomain_id = 1
    complement_subdomain_id = 0
    execute_on = 'INITIAL'
  []
[]


[Variables]
    [u]
    []
[]

#Try use the auxillary variable in the adaptivity criterion instead of assigning it to the displacement variable
[ICs]
    [u_ic]
      type = FunctionIC
      function = image_func
      variable = u
    []
[]

[Problem]
    type = FEProblem
    solve = false  # that was the other way no to solve the problem, but only generate the mesh
[]

[Executioner]
    type = Transient
    num_steps = 1
    dt = 0.1
[]

[Adaptivity]
    max_h_level = 10
    initial_steps = 5
    initial_marker = marker
    [Indicators]
      [indicator]
        type = GradientJumpIndicator
        variable = auxBlockID
      []
    []
    [Markers]
      [marker]
        type = ErrorFractionMarker
        indicator = indicator
        refine = 0.9
      []
    []
[]

[Outputs]
    exodus = true
[]
```

**Reply by GiudGiud** [reply]
hello

first, I dont think the ImageFunction worked as ewxpected. You should show auxBlockID instead of object_ID in the paraview output to check
the adaptivity will have to be figured out next

**Reply by Rahim-Habibi** [reply]
Thnaks @GiudGiud 

here is auxBlockID in paraview

 
![OIP-paraview](https://github.com/user-attachments/assets/c61940eb-0e52-47e0-8e40-aa2834782bec)

**Reply by GiudGiud** [reply]
Are you plotting those at the end of the simulation or after a time step? Parlaview will plot the initial state by default

**Reply by Rahim-Habibi** [reply]
@GiudGiud  it is at the end of simulation, the simulation time is just 0.1s.
[Executioner]
    type = Transient
    num_steps = 1
    dt = 0.1
[]

**Reply by GiudGiud** [reply]
in paraview did you change the time to select the end of the simulation.
in MOOSE you should be fine, we output all time steps to the exodus file

**Reply by Rahim-Habibi** [reply]
@GiudGiud,
yes i selected the end of simulation in paraview. 
![OIP-paraview](https://github.com/user-attachments/assets/0c56bf69-589f-4a2f-aeb1-c868ee59da19)

**Reply by GiudGiud** [reply]
ok then ImageFunction isnt working as expected here
